### PR TITLE
Add consistency checks in map and array vectors

### PR DESCRIPTION
**Pull request by bikramSingh91**
This adds consistency checks to both map and array vectors that
ensure that the sizes of the offset and size buffers are valid.
Moreover, adds a more through check that only fires in debug
builds which checks whether the values in the offset and size
buffers are valid.

### CODE REVIEW

**Comment by bikramSingh91** [comment]
fixes part of #3043

File: velox/vector/ComplexVector.h [hunk 1]
```diff
@@ -276,7 +276,11 @@ struct ArrayVectorBase : BaseVector {
         offsets_(std::move(offsets)),
         rawOffsets_(offsets_->as<vector_size_t>()),
         sizes_(std::move(lengths)),
-        rawSizes_(sizes_->as<vector_size_t>()) {}
+        rawSizes_(sizes_->as<vector_size_t>()) {
+    VELOX_CHECK_GE(
```
Comment (mbasmanova) [inline velox/vector/ComplexVector.h:280]: Nice.

File: velox/vector/ComplexVector.h [hunk 2]
```diff
@@ -286,6 +290,38 @@ struct ArrayVectorBase : BaseVector {
       VectorPtr* targetKeys,
       const BaseVector* sourceKeys);
 
+  /// Checks whether offsets and sizes are valid values and point to valid
+  /// indices in the child vectors(elements for array, keys and values for
+  /// map.
+  void checkConsistency(size_t childVectorSize) {
+#ifndef NDEBUG
+    for (auto i = 0; i < BaseVector::length_; ++i) {
+      const bool isNull =
```
Comment (mbasmanova) [inline velox/vector/ComplexVector.h:299]: nit: since this is debug only, why not just use isNullAt(i)?

**Review by mbasmanova — APPROVED** [review]
@bikramSingh91 Nice. Thank you!

**Comment by Yuhta** [comment]
The change is good, it's just `Buffer::size` is often ignored in most places, so there might be a few places need to be fixed.

**Comment by bikramSingh91** [comment]
Looks like 184 tests failed and most hit the consistency check verifying offset and size validity. Will looking this and if the consistency check brings up more things to fix, I'll move it out into a separate PR

**Comment by Yuhta** [comment]
@bikramSingh91 I am wondering if we should move `size` to `NonPODAlignedBuffer` only, since it does not make much sense to keep both capacity and size for POD buffers.

**Comment by bikramSingh91** [comment]
all failures are due to the additional check i added for the actual values in offset and size in checkConsistency(). Moving it out of this PR to get this merged

**Comment by bikramSingh91** [comment]
> @bikramSingh91 I am wondering if we should move `size` to `NonPODAlignedBuffer` only, since it does not make much sense to keep both capacity and size for POD buffers.

@Yuhta I agree we should clean up these differences and have a more consistent usage patter. Recently ran into another issue related to this and filed an issue #3089 to track this

**Review by xiaoxmeng — APPROVED** [review]
LGTM. Thanks!

**Pull request closed without merging**